src/main/downloadsManager.ts

```diff
@@ -374,6 +378,46 @@ export class DownloadsManager extends JsonFileManager<DownloadedItems> {
         this.saveAll(downloads);
     };
 
+    private handleSelectDownload = (event: IpcMainInvokeEvent, startFrom: string) => {
+        return this.selectDefaultDownloadDirectory(
+            startFrom,
+            localizeMessage('main.downloadsManager.specifyDownloadsFolder', 'Specify the folder where files will download'),
+        );
+    }
+
+    private selectDefaultDownloadDirectory = async (startFrom: string, message: string) => {
+        log.debug('handleSelectDownload', startFrom);
+
+        const result = await dialog.showOpenDialog({defaultPath: startFrom || Config.downloadLocation,
+            message,
+            properties:
+        ['openDirectory', 'createDirectory', 'dontAddToRecent', 'promptToCreate']});
+        return result.filePaths[0];
+    }
+
+    private verifyMacAppStoreDownloadFolder = async (fileName: string) => {
+        let downloadLocation = Config.downloadLocation;
+
+        // eslint-disable-next-line no-undef
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        if (__IS_MAC_APP_STORE__ && downloadLocation) {
+            try {
+                const savePath = this.getSavePath(downloadLocation, fileName);
+                fs.writeFileSync(savePath, '');
+                fs.unlinkSync(savePath);
+            } catch (e) {
+                downloadLocation = await this.selectDefaultDownloadDirectory(
```

Comment:
Should this still save the file after the path has been selected? Or maybe I'm misunderstanding this and the `writeFileSync` in the `try` block doesn't actually save the file

Reply:
Yeah this is a weird case. Because of the async nature of this code, after the method finishes for some reason the `DownloadItem` is marked destroyed and we can't re use it. So for the cases where we need to pick a folder to save to, we have this `willDownloadUrls` queue that we add to to recreate downloads effectively.